### ISSUE TITLE
'Matching events' when using 'Element matches HTML selector' in Actions lies to users

### ISSUE THREAD

**Issue by bretthoerner** [PostHog/posthog#20492]
The string you put in the `Element matches HTML selector` section of actions is just searched among the elements chain in ClickHouse. This makes sense! Parsing and selecting based on CSS at read time across all events in CH would be too slow.

The problem is that it lies to users, they see `Matching events` and get confused when their action doesn't fire (often they are expecting a Webhook send).

![image](https://github.com/PostHog/posthog/assets/1752/78232398-a10d-4365-b84c-1ccdf37e499f)

In the example above, the user entered `foo`, thinking it would hit elements with the `foo` class in their HTML. They see matches, are happy, and hit save. The problem is that a plain string CSS selector searches for *elements*, i.e. `<foo>`, to select a class they needed to put `.foo`. Nothing will actually match and the action won't fire.

Our CSS selector logic [is already implemented in TypeScript in the plugin-server](https://github.com/PostHog/posthog/blob/72b7077ecf497ef231c67a009776c40fd34d5f0d/plugin-server/src/worker/ingestion/action-matcher.ts#L470). What if we shared the code with the frontend and filtered out matching events on the front-end? It wouldn't be perfect, but it would avoid these common false positives.

Alternatively, or in addition to, we could have a very loud message/banner near `Matching events` when the CSS selector is non-empty.

**Comment by mariusandra** [comment]
I believe the "correct" fix here is to match directly against the elements chain in the plugin server using literally the same regex and the same regex engine (re2).

This _is_ already implemented here https://github.com/PostHog/posthog/pull/17042 ... but I keep finding it an uphill battle to even have anyone acknowledge this PR's existence, nor to find time to push for it myself... :/

**Comment by bretthoerner** [comment]
@mariusandra Sorry for the dumb question, that PR predates my time here. Does "using literally the same regex and the same regex engine (re2)" mean it would match what the frontend query is doing *today*? Or would this change the behavior of both the frontend query and the plugin-server?

I ask because my `foo` example above shows how the current front end query is incorrect.

**Comment by mariusandra** [comment]
Having looked into this a bit, the problem is that the "tag_name" selector here isn't bounded: https://github.com/PostHog/posthog/blob/10cd61d89b97714e0cf9d868f66e6ccb60537e65/posthog/models/property/util.py#L850-L854

Fixing this might be complicated due to `re2` that ClickHouse uses for regex matching (no support for lookbehinds). It might be possible/easy some other way (just `/(^|[^a-z])/` before the tag name might do the trick?), but someone needs to dig in.

Due to ClickHouse we can't not use regex-s and `re2` for this matching. Even if we're not 100% compliant to the CSS standard, if we're uniformly broken, that's probably good enough. Thus I think the best solution for users is for us to focus on making the queries and webhook matching work the same way.

The easiest solution I can think of (save from completing the HogVM PR which gets the regex from python directly), is to re-implement the regex code from [here](https://github.com/PostHog/posthog/blob/10cd61d89b97714e0cf9d868f66e6ccb60537e65/posthog/models/property/util.py#L847-L864) and [here](https://github.com/PostHog/posthog/blob/2a222ad740fed1e8051ecee4cfa0761c90408d83/posthog/models/event/event.py#L80-L120) in JS for the plugin server, and replace the lines [here](https://github.com/PostHog/posthog/blob/1f6ce77e66dd8771a25315d7aca03add62567a40/plugin-server/src/worker/ingestion/action-matcher.ts#L270-L272) to instead match against the `elements_chain`, which may need to be passed in from somewhere.

**Comment by mariusandra** [comment]
> @mariusandra Sorry for the dumb question, that PR predates my time here. Does "using literally the same regex and the same regex engine (re2)" mean it would match what the frontend query is doing _today_? Or would this change the behavior of both the frontend query and the plugin-server?
> 
> I ask because my `foo` example above shows how the current front end query is incorrect.

Regarding "re2", there are two regex "systems" that I'm aware of, "re2" and "perl style regex-s". ClickHouse uses re2 due to its constant time guarantees, but most programming languages implement the perl version.

The current HogVM implementation uses "whatever is available", and 99% of the time they work the same way, but there are some differences in syntax with regards to query modifiers. The fix here is to just add a package, `import { match } from 're2'`, and use that instead of the built in functions.

Regarding using the same regex, my previous comment touched upon how we can do that.

The "magic" with HogVM however is that each action gets compiled in python into "bytecode", including inlining the elements_chain regex in that, and then this is passed to the plugin server. It's already there: look at the `bytecode` field in postgres for `posthog_action`.

**Comment by bretthoerner** [comment]
It sounds like whatever we do we should also clarify that the HTML/CSS Selector logic isn't accurate.

**Comment by mariusandra** [comment]
We do say that not everything is supported, and describe what is supported here: https://posthog.com/docs/data/actions#matching-selectors 

For me the discussion here is about a bug that's caused by us having two incompatible implementations of the same logic. We should fix the discrepancy, and [this change](https://github.com/PostHog/posthog.com/pull/7886) in docs feels like enough to guide those users who do read docs towards the right place.

**Comment by bretthoerner** [comment]
Maybe it's a nitpick, but the example in the ticket (which has come up in support) is a mismatch on the query side and not a "things aren't supported." But maybe that can be fixed in the regex (not sure if that's what you were saying [above](https://github.com/PostHog/posthog/issues/20492#issuecomment-1959553352) but you replied right when I did so it seems unrelated?).

I'm not familiar with HogVM, I think what I don't understand still is why the same regex we pass to ClickHouse can't be used in Python and Node via `re2` bindings (without anything fancier than literally reusing the same regex).

**Comment by mariusandra** [comment]
1. Yes, it's wrong for the selector `"foo"` to match an element with the class `foo`. It should only match the tag name "foo". Our regex is built wrong, and it should be fixed. [At the top of this comment](https://github.com/PostHog/posthog/issues/20492#issuecomment-1959553352) I described where that fix should go, and why it currently works as it does.
2. We don't have the "selector to regex" code in node right now, only in python. We need to build/port it. I described at the bottom of the same comment where the relevant code is in python, and where it should be placed in the plugin server.
3. Even though the CSS selector itself is wrong (point 1), the bigger problem is that it's inconsistent in the two places where we use it. This is not the only inconsistency. If we make it wrong _and_ consistent, it would solve a lot of problems. Making it correct (point 1) is also a great idea, but there will be some other inconsistency next month. Thus the "better" solution in my mind is to do point 2. Modifying the regex itself (point 1) might have side effects we're not aware of, and will require more work/testing to roll out.
4. All that said, there is already a system that generates the regex in python, and directly passes it to node for evaluation in the form of HogVM bytecode. We convert _every_ action including all filters into a compact bytecode format... which isn't used anywhere. The node HogVM implementation ([literally ~200 lines of javascript](https://github.com/PostHog/posthog/blob/558b3b0e304468c12f2cdafd7c94c8f1a057e0fa/hogvm/typescript/src/bytecode.ts)) is just missing re2 for 100% compatibility. It's easy to add. Sadly I'm met with a lot of resistance every time I bring up HogVM. In my mind it massively simplifies the action matching, and paves way for a fully-rust plugin server if ported ([plus other goodies](https://github.com/PostHog/product-internal/pull/494)), but I seem to have trouble getting the message across.
